router/request.go

```diff
@@ -523,6 +535,165 @@ func (h *Handlers) DeleteComment(c echo.Context) error {
+	}
+	if req.Comment == "" {
+		if !IsAbleNoCommentChangeStatus(req.Status, request.Status) {
+			message := fmt.Sprintf("unable to change %v to %v without comment", request.Status.String(), req.Status.String())
+			c.Logger().Error(message)
+			return echo.NewHTTPError(http.StatusBadRequest, errors.New(message))
+		}
+	}
+
+	u, err := h.Repository.GetUserByID(ctx, user.ID)
+	if err != nil {
+		if ent.IsNotFound(err) {
+			c.Logger().Error(err)
+			return echo.NewHTTPError(http.StatusNotFound, err)
+		}
+		c.Logger().Error(err)
+		return echo.NewHTTPError(http.StatusInternalServerError, err)
+	}
+	if u.Admin {
+		if !IsAbleAdminChangeState(req.Status, request.Status) {
+			message := fmt.Sprintf("admin unable to change %v to %v", request.Status.String(), req.Status.String())
+			c.Logger().Error(message)
+			return echo.NewHTTPError(http.StatusBadRequest, errors.New(message))
+		}
+		if req.Status == model.Submitted && request.Status == model.Accepted {
+			targets, err := h.Repository.GetRequestTargets(ctx, requestID)
+			if err != nil {
+				c.Logger().Error(err)
+				return echo.NewHTTPError(http.StatusInternalServerError, err)
+			}
+			var paid bool
+			for _, target := range targets {
+				if target.PaidAt != nil {
+					paid = true
+					break
+				}
+			}
+			if paid {
+				c.Logger().Error("someone already paid")
+				return echo.NewHTTPError(http.StatusBadRequest, errors.New("someone already paid"))
+			}
+		}
+	}
+
+	if !u.Admin && user.ID == request.CreatedBy {
+		if !IsAbleCreatorChangeStatus(req.Status, request.Status) {
+			message := fmt.Sprintf("creator unable to change %v to %v", request.Status.String(), req.Status.String())
+			c.Logger().Error(message)
+			return echo.NewHTTPError(http.StatusBadRequest, errors.New(message))
+		}
+	}
+
+	if user.ID != request.CreatedBy && !u.Admin {
+		return echo.NewHTTPError(http.StatusUnauthorized)
+	}
+
+	// create status and comment
+	created, err := h.Repository.CreateStatus(ctx, requestID, user.ID, req.Status)
+	if err != nil {
+		c.Logger().Error(err)
+		return echo.NewHTTPError(http.StatusInternalServerError, err)
+	}
+	var resComment string
+	if req.Comment != "" {
+		comment, err := h.Repository.CreateComment(ctx, req.Comment, request.ID, user.ID)
+		if err != nil {
+			c.Logger().Error(err)
+			return echo.NewHTTPError(http.StatusInternalServerError, err)
+		}
+		resComment = comment.Comment
+	}
+
+	res := &Status{
+		CreatedBy: user.ID,
+		Status:    created.Status,
+		Comment:   resComment,
+		CreatedAt: created.CreatedAt,
+	}
+
+	return c.JSON(http.StatusOK, res)
+}
+
+func IsAbleNoCommentChangeStatus(status, latestStatus model.Status) bool {
+	if status == model.FixRequired && latestStatus == model.Submitted {
+		return false
+	}
+	if status == model.Rejected && latestStatus == model.Submitted {
+		return false
+	}
+	if status == model.Submitted && latestStatus == model.Accepted {
+		return false
+	}
+	return true
```

Comment:
```suggestion
	if status == model.FixRequired && latestStatus == model.Submitted ||
	   status == model.Rejected && latestStatus == model.Submitted ||
	   status == model.Submitted && latestStatus == model.Accepted {
		return false
	}
	return true
```

昔俺が書いたコードだけど、こういうほうが分かりやすいかも？